### PR TITLE
Add virtual dtor to algo_base

### PR DESCRIPTION
**Pull request by matze**
Otherwise, compilers will complain if verifier is used as a member variable.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
which compiler is that? typically as long as there's one virtual member they don't complain?

**Comment by matze** [comment]
> which compiler is that?

I only tested with Clang 9 and `-Wall`.

> typically as long as there's one virtual member they don't complain?

I don't really understand that question. But here's a MWP that outlines the issue

```cpp
struct Foo {
    jwt::verifier<jwt::default_clock, jwt::picojson_traits> verifier{
        jwt::verify<jwt::default_clock, jwt::picojson_traits>({})
            .allow_algorithm(jwt::algorithm::hs256{"secret"})
    };
};
```
Note, that there is no warning if I do not call `allow_algorithm()`.

**Comment by Thalhammer** [comment]
I think this might actually be a false positive. shared_ptr contains a type erased destructor and since the full type is used in make_shared it should in theory call the right thing.

That said I dont think adding a virtual destructor should be an issue.

**Comment by prince-chrismc** [comment]
Welp! I found this, https://stackoverflow.com/a/461224/8480874, which allude to the fact the default is non-virtual and after checking https://en.cppreference.com/w/cpp/language/destructor the implicitly declared is not required to be virtual.

So this is certainly required. 

You learn something new everyday =D I just assumed compilers followed "public and virtual or protected and nonvirtual" but i guess not!

thank you for your contribution!

**Comment by Thalhammer** [comment]
> https://en.cppreference.com/w/cpp/language/destructor the implicitly declared is not required to be virtual.

I know and most of the time it is not.

> So this is certainly required.

No it should not.
It would only be needed if you delete a class using a pointer to a base class. In our case this would mean deleting algo<T> via a algo_base* pointer.
However we never do this as std::shared_ptr is required to contain a deleter, which defaults to std::default_delete<T> which in our case (when created) is std::default_delete<algo<T>>.
I.e. if the shared_ptr is destroyed it should call something which is in essence `delete static_cast<algo<T>*>(ptr);`. Thus a virtual destructor would not be required.

TLDR: I 100% agree we should merge this (just ot be safe), but in my opinion this is a compiler/library defect.

@matze Whats the exact warning you get ?

**Comment by matze** [comment]
```
matthias@ubuntu:~/dev/jwt-cpp (master %) clang++-10 -Wall test.cpp -Iinclude
In file included from test.cpp:1:
In file included from include/jwt-cpp/jwt.h:8:
In file included from include/picojson/picojson.h:36:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/iostream:39:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ostream:38:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ios:42:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/ios_base.h:41:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/locale_classes.h:40:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/string:41:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/allocator.h:46:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/x86_64-linux-gnu/c++/9/bits/c++allocator.h:33:
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ext/new_allocator.h:153:4: warning: destructor called on non-final 'jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>' that has virtual functions but non-virtual destructor [-Wdelete-non-abstract-non-virtual-dtor]
        { __p->~_Up(); }
          ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/alloc_traits.h:497:8: note: in instantiation of function template specialization '__gnu_cxx::new_allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >::destroy<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >' requested here
        { __a.destroy(__p); }
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:557:28: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> > >::destroy<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >' requested here
        allocator_traits<_Alloc>::destroy(_M_impl._M_alloc(), _M_ptr());
                                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:543:2: note: in instantiation of member function 'std::_Sp_counted_ptr_inplace<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, __gnu_cxx::_S_atomic>::_M_dispose' requested here
        _Sp_counted_ptr_inplace(_Alloc __a, _Args&&... __args)
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:680:6: note: in instantiation of function template specialization 'std::_Sp_counted_ptr_inplace<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, __gnu_cxx::_S_atomic>::_Sp_counted_ptr_inplace<jwt::algorithm::hs256 &>' requested here
            _Sp_cp_type(__a._M_a, std::forward<_Args>(__args)...);
            ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:1344:14: note: in instantiation of function template specialization 'std::__shared_count<__gnu_cxx::_S_atomic>::__shared_count<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, jwt::algorithm::hs256 &>' requested here
        : _M_ptr(), _M_refcount(_M_ptr, __tag, std::forward<_Args>(__args)...)
                    ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr.h:359:4: note: in instantiation of function template specialization 'std::__shared_ptr<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, __gnu_cxx::_S_atomic>::__shared_ptr<std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, jwt::algorithm::hs256 &>' requested here
        : __shared_ptr<_Tp>(__tag, std::forward<_Args>(__args)...)
          ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr.h:701:14: note: in instantiation of function template specialization 'std::shared_ptr<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >::shared_ptr<std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, jwt::algorithm::hs256 &>' requested here
      return shared_ptr<_Tp>(_Sp_alloc_shared_tag<_Alloc>{__a},
             ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr.h:717:19: note: in instantiation of function template specialization 'std::allocate_shared<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, std::allocator<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256> >, jwt::algorithm::hs256 &>' requested here
      return std::allocate_shared<_Tp>(std::allocator<_Tp_nc>(),
                  ^
include/jwt-cpp/jwt.h:1799:28: note: in instantiation of function template specialization 'std::make_shared<jwt::verifier<jwt::default_clock, jwt::picojson_traits>::algo<jwt::algorithm::hs256>, jwt::algorithm::hs256 &>' requested here
                        algs[alg.name()] = std::make_shared<algo<Algorithm>>(alg);
                                                ^
test.cpp:6:14: note: in instantiation of function template specialization 'jwt::verifier<jwt::default_clock, jwt::picojson_traits>::allow_algorithm<jwt::algorithm::hs256>' requested here
            .allow_algorithm(jwt::algorithm::hs256{"secret"})
             ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/ext/new_allocator.h:153:10: note: qualify call to silence this warning
        { __p->~_Up(); }
                ^
1 warning generated.
```

Hmm, seems the stdlib comes from gcc …

**Comment by prince-chrismc** [comment]
https://godbolt.org/z/sxqcWo

No such luck reproducing on GCC
